c7n/actions.py

```diff
@@ -238,60 +327,30 @@ def get_groups(self, resources, metadata_key=None):
-                # rds resource security groups
-                if metadata_key and isinstance(r['VpcSecurityGroups'][0], dict):
-                    rgroups = [g[metadata_key] for g in r['VpcSecurityGroups']]
-                else:
-                    rgroups = [g for g in r['VpcSecurityGroups']]
-            elif r.get('VPCOptions', {}).get('SecurityGroupIds', []):
-                # elasticsearch resource security groups
-                if metadata_key and isinstance(
-                        r['VPCOptions']['SecurityGroupIds'][0], dict):
-                    rgroups = [g[metadata_key] for g in r[
-                        'VPCOptions']['SecurityGroupIds']]
-                else:
-                    rgroups = [g for g in r['VPCOptions']['SecurityGroupIds']]
-            # use as substitution for 'Groups' or '[Vpc]SecurityGroups'
-            # unsure if necessary - defer to coverage report
-            elif metadata_key and r.get(metadata_key):
-                rgroups = [g for g in r[metadata_key]]
+        if isinstance(add_target_group_ids, list):
+            add_names = [name for name in add_target_group_ids if not name.startswith('sg-')]
+            add_target_group_ids += self.resolve_security_group_names(add_names)
+        elif isinstance(add_target_group_ids, str) and not add_target_group_ids.startswith('sg-'):
+            # Can assume sg's won't start with 'sg-'
+            #
+            # The name of the security group.
+            #    Constraints: Up to 255 characters in length. Cannot start with sg- .
+            #
+            # https://docs.aws.amazon.com/cli/latest/reference/ec2/create-security-group.html
 
-            # Parse remove_groups
-            if remove_target_group_ids == 'matched':
-                remove_groups = r.get('c7n:matched-security-groups', ())
-            elif remove_target_group_ids == 'network-location':
-                for reason in r.get('c7n:NetworkLocation', ()):
-                    if reason['reason'] == 'SecurityGroupMismatch':
-                        remove_groups = list(reason['security-groups'])
-            elif remove_target_group_ids == 'all':
-                remove_groups = rgroups
-            elif isinstance(remove_target_group_ids, list):
-                remove_groups = remove_target_group_ids
-            elif isinstance(remove_target_group_ids, six.string_types):
-                remove_groups = [remove_target_group_ids]
-
-            # Parse add_groups
-            if isinstance(add_target_group_ids, list):
-                add_groups = add_target_group_ids
-            elif isinstance(add_target_group_ids, six.string_types):
-                add_groups = [add_target_group_ids]
+            add_target_group_ids = self.resolve_security_group_names([add_target_group_ids])
+
+        if isinstance(remove_target_group_ids, list):
+            remove_names = [name for name in remove_target_group_ids if not name.startswith('sg-')]
```

Comment:
should this check just be in resolve_security_group_names, ie then its just resovle_security_groups and you can use it for add/remove clauses without the additional prefix logic, which would localize that logic to the resolve function.